### PR TITLE
echo mtp

### PR DESCRIPTION
**Pull request by FeixLiu**
### PR Category

Distributed Strategy

### PR Types

New features

### Description

echo mtp

### 是否引起精度变化

否

### CODE REVIEW

File: python/paddle/distributed/fleet/meta_parallel/parallel_layers/pp_layers.py [hunk 1]
```diff
@@ -963,6 +981,50 @@ def _build_layer(self):
...
+        share the same parameter memory. Asserts that all parameters are successfully
+        aliased.
+
+        Args:
+            layer: Destination layer whose transformer_layer parameters will be replaced.
+            src_layer: Source layer providing the shared parameters.
+        """
+        source_params = dict(src_layer.named_parameters())
+
+        dest_module = dest_layer.transformer_layer
```
Comment (risemeup1111) [inline python/paddle/distributed/fleet/meta_parallel/parallel_layers/pp_layers.py:998]: P2 **优先级：P2**

`shared_submodule_weight_only` 暴露在通用的 `SharedLayerDesc` 上，但这里把目标子层硬编码成 `dest_layer.transformer_layer`，并且后续用 `src_layer.named_parameters()` 作为源参数集合。这样只有 MTP 里恰好命名为 `transformer_layer` 的结构能工作；如果用户通过 `shared_weight_attr` 暴露的是 `decoder_weights = self.decoder.named_parameters()` 这类同样合法的共享子模块，会在这里直接 `AttributeError`，或者因为 helper 忽略 `shared_weight_attr` 而别名了不该共享的参数。

建议让 helper 以 `shared_weight_attr` 展开的参数集合为准，或者新增显式的子模块 attr 参数并在构造时校验；不要在框架层固定 `transformer_layer` 这个业务字段。示例形态：

```python
def _alias_shared_layer(self, dest_layer, src_layer, weight_attr, submodule_attr):
    source_params = dict(_iter_shared_params(src_layer, weight_attr))
    dest_module = getattr(dest_layer, submodule_attr)
    for param_name, dest_param in dest_module.named_parameters():
        source_param = source_params.get(param_name)
        ...
```

File: python/paddle/distributed/fleet/meta_parallel/parallel_layers/pp_layers.py [hunk 2]
```diff
@@ -1010,32 +1072,58 @@ def flush_into_run_function():
...
+                        self.add_sublayer(str(layer_index), instance)
+                        self._alias_shared_layer(
+                            instance, self.shared_layers[layer.layer_name]
                         )
-                        param.is_firstly_shared = True
-
-                if layer.forward_func is None:
-                    run_function.append(self.shared_layers[layer.layer_name])
-
+                    run_function.append(instance)
```
Comment (risemeup1111) [inline python/paddle/distributed/fleet/meta_parallel/parallel_layers/pp_layers.py:1091]: P1 **优先级：P1**

`shared_submodule_weight_only=True` 分支绕过了已有 `SharedLayerDesc.forward_func` 语义。老分支在 `forward_func` 非空时会 append `partial(layer.forward_func, shared_layer)`，但这里无条件 append `instance`，所以组合 `SharedLayerDesc(..., forward_func=..., shared_submodule_weight_only=True)` 时会静默执行整个 layer 的 `forward()`，而不是用户指定的 helper；这会改变现有 `SharedLayerDesc` 参数组合的行为，MTP/logits helper 这类场景会得到错误前向。

建议按老逻辑处理 `forward_func`，并在 interleave chunk 场景继续把实际 layer 注册进 `PipelineLayerChunk`，例如：

```python
if layer.forward_func is None:
    run_function.append(instance)
else:
    run_function.append(partial(layer.forward_func, instance))
    if self._num_virtual_pipeline_stages > 1:
        run_function.add_sublayer(layer.layer_name, instance)
```

请同时补一个 `shared_submodule_weight_only=True` 与 `forward_func` 组合的测试，避免后续再次退化。

File: python/paddle/distributed/fleet/meta_parallel/parallel_layers/pp_layers.py [hunk 3]
```diff
@@ -740,65 +742,81 @@ def _synchronize_shared_weights(self):
         for key, comm in self.shared_comm.items():
             with paddle.framework.no_grad():
                 for weight_attr in comm['weight_attr']:
-                    paddle.distributed.broadcast(
-                        getattr(comm['layer'], weight_attr),
-                        src=min(comm['ranks']),
-                        group=comm['group'],
-                    )
+                    obj = getattr(comm['layer'], weight_attr)
```
Comment (risemeup1111) [inline python/paddle/distributed/fleet/meta_parallel/parallel_layers/pp_layers.py:745]: P1 **优先级：P1**

这里已经允许 `shared_weight_attr` 返回 `named_parameters()` 这类非 Tensor 对象，但 `_construct_shared_comm()` 里 `sync_param/sync_moment` 的打标路径仍然对 `getattr(..., weight_attr)` 直接设置 `.color`。当用户开启现有的 `pp_configs.sync_param=True` 或 `sync_moment=True` 时，新增测试里的 `transformer_layer_weights` 会返回 generator，构建通信组时会对 generator 设置属性并触发 `AttributeError`；即使返回 list，优化器 `_pp_filter_fn` 也只能筛到真正带 `color` 的参数对象。

建议把“共享 attr 展开为参数”的逻辑抽成一个 helper，并在 broadcast、grad allreduce、`is_firstly_shared`、`color` 打标四处复用；`shared_weight_name` 也需要包含子参数名，避免多个参数都使用同一个排序键。修复形态可以类似：

```python
def _iter_shared_params(layer, weight_attr):
    obj = getattr(layer, weight_attr)
    if isinstance(obj, paddle.Tensor):
        yield weight_attr, obj
    else:
        yield from obj

for weight_attr in shared_attrs:
    for name, param in _iter_shared_params(self.shared_layers[layer_name], weight_attr):
        param.color = {
            "color": f"{SHARED_WEIGHT_SYNC_PREFIX}_{comm_key}",
            "group": hcg.get_sharding_parallel_group(),
            "shared_weight_name": f"{weight_attr}.{name}" if name else weight_attr,
            "broadcast_group": group,
        }
```

同时请补一个 `shared_submodule_weight_only=True` 且开启 `sync_param/sync_moment` 的回归测试，覆盖这条配置组合。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
发现几处需要在合入前处理的问题，细节已放在行级评论中。

- P3 **优先级：P3** 非行级：PR 标题和 Description 目前只有 `echo mtp`，无法说明变更范围、风险和验证方式。建议标题改为 `[Distributed Strategy] 支持 SharedLayerDesc 共享子模块权重`，Description 补充本次新增的 `shared_submodule_weight_only` 行为、参数/动量同步影响，以及已覆盖的测试场景。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: test/legacy_test/test_shared_submodule_weight_only_unit.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+
+
+class TestBuildLayerTraditionalNonTensorPath(unittest.TestCase):
+    """Cover traditional SharedLayerDesc path with non-Tensor weight_attr."""
+
+    def test_is_firstly_shared_non_tensor_traditional(self):
+        """Simulate _build_layer_impl traditional path: non-Tensor obj marks all params."""
+        layer = SimpleTransformerLayer()
+        # This is the exact logic in _build_layer_impl lines 1097-1106
+        for weight_attr in ['transformer_layer_weights']:
```
Comment (risemeup1111) [inline test/legacy_test/test_shared_submodule_weight_only_unit.py:259]: P2 **优先级：P2**

这个用例把 `_build_layer_impl` 中给非 Tensor `shared_weight_attr` 设置 `is_firstly_shared` 的实现直接复制到测试里了，实际没有调用生产代码。这样即使 `_build_layer_impl` 的对应分支回归，测试仍会通过，也不能覆盖 `pp_layers.py` 里新增的目标行。

请改成通过最小 `PipelineLayer`/mock 对象执行 `_build_layer_impl`，或放到已有 pipeline 集成测试中断言实际构建后的参数标记。形态可以类似：

```python
pipe = object.__new__(PipelineLayer)
pipe._num_virtual_pipeline_stages = 1
pipe._use_dualpipev = False
pipe.run_function = []
pipe.shared_layers = paddle.nn.LayerDict()
pipe._layers_desc = [
    SharedLayerDesc(
        'shared',
        SimpleTransformerLayer,
        shared_weight_attr='transformer_layer_weights',
    )
]
# 初始化 _build_layer_impl 需要的其余字段后调用真实实现
PipelineLayer._build_layer_impl(pipe, 0, 1)
for _, param in pipe.shared_layers['shared'].named_parameters():
    self.assertTrue(param.is_firstly_shared)
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已看新增提交，主要是补充单测；核心实现没有变化。前一轮的阻塞问题仍由已有行级线程跟踪，本次只新增了一条关于单测覆盖方式的行级建议。

- P3 **优先级：P3** 非行级：PR 标题和 Description 仍是 `echo mtp`，建议改为 `[Distributed Strategy] 支持 SharedLayerDesc 共享子模块权重`，并在描述中补充 `shared_submodule_weight_only` 的行为、同步配置影响和验证范围。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by LiYuRio — APPROVED** [review]
